65: Attach Library used in Pattern Lab

evanmwillhite opened the pull request:
This adds support for Pattern Lab to use attach_library instead of using a global JavaScript file. It adds the scripts inline into the component via the existing Twig function. This will allow us to remove the styleguide-scripts gulp task - see https://github.com/fourkitchens/emulsify-gulp/issues/65. Here's the companion PR for this task in Emulsify Gulp: https://github.com/fourkitchens/emulsify-gulp/pull/77

**Adds, Changes**:
- Changes `attach_library` function from a passthrough to a useful function for Pattern Lab, accepting the library name using the same syntax as Drupal and searching for the correct file in the `dist` directory by that name.
- Changes relevant component files to use that attach_library function (and removes from Drupal tpl)
- Removes styleguide-scripts file reference from PL footer
- Random: Adds minified SVG file from https://github.com/fourkitchens/emulsify-gulp/pull/76.

**To test:**
- [x] Checkout this branch in your Emulsify instance
- [x] If you are using an existing Emulsify instance, remove vendor and node_modules directories then run `composer install` and `npm/yarn install`
- [x] Use the [`65-remove-styleguide-scripts` branch](https://github.com/fourkitchens/emulsify-gulp/pull/77) in Emulsify Gulp
- [x] Run `yarn start` and ensure no errors
- [ ] View any existing components using JS ([accordion](http://localhost:3000/pattern-lab/public/?p=viewall-molecules-accordion-item), [menu](http://localhost:3000/pattern-lab/public/?p=molecules-main-menu)) and ensure they still work
- [ ] View the menu component in a local Drupal instance using Emulsify and verify it is still loading correctly.

**Questions**
- ~Is there a better way to handle [this](https://github.com/fourkitchens/emulsify/blob/35f3ec71ddc4609cf05387b5b4f2d4bb49af1028/components/_twig-components/functions/pl_attach-library.function.php#L10-L13) for recursive searches?~

ModulesUnraveled commented:
I think @ccjjmartin should review code on this. At least the PHP stuff that you had a question about.

ccjjmartin commented on components/_twig-components/functions/pl_attach-library.function.php:
```diff
@@ -5,5 +5,20 @@
-  return $string;
-});
+  // Get Library Name from string.
+  $libraryName = substr($string, strpos($string, "/") + 1);
```
This could return FALSE.

evanmwillhite replied:
when the library name is incorrect, it just doesn't load the file so I'm OK with this.

ccjjmartin commented on components/_twig-components/functions/pl_attach-library.function.php:
```diff
@@ -5,5 +5,20 @@
+  $libraryName = substr($string, strpos($string, "/") + 1);
+
+  // Search dist root directory for library.
+  $files = glob('dist/*/' . $libraryName . '/' . $libraryName . '*.js');
```
Do we care about the situations where $libraryName is different from the file name?

evanmwillhite replied:
actually, this code needed to be refactored. The file name is covered by `*.js` so the library name just has to match the name of the component.

ccjjmartin commented on components/_twig-components/functions/pl_attach-library.function.php:
```diff
@@ -5,5 +5,20 @@
+  $files = glob('dist/*/' . $libraryName . '/' . $libraryName . '*.js');
+  // Search dist nested directory for library.
+  if (empty($files)) {
+    $files = glob('dist/*/*/' . $libraryName . '/' . $libraryName . '*.js');
```
Should we be using the dist directory here or the source directory?

evanmwillhite replied:
dist is correct - it's where the final minified file from the gulp scripts task is placed (and then a map file is created for debugging as well)

ccjjmartin commented:
@evanmwillhite Looks like people are commonly complaining about glob not supporting recursive searches on php.net: http://php.net/manual/en/function.glob.php  You solution looks cleaner than some of the examples there.

evanmwillhite commented:
Hmm.. ok. Glad to hear I didn't go off the rails, but sad to hear there's not a better solution. I guess we could just document it well. Currently, it is such that:

1. Your JS file has to be in a component either 1 or 2 deep (e.g., `atoms/COMPONENT/component.js` or `molecules/DIRECTORY/COMPONENT/component.js`
2. You have to use the component name (COMPONENT in the examples above) in your attach_library function.

That too restrictive?

evanmwillhite commented:
The more I think about this, the more I think there may need to be more parity between Drupal's implementation and Pattern Lab's. For instance, if there is a library that has dependencies on the Drupal side, the only way to handle that on Pattern Lab's side cleanly would be to revert to throwing those dependencies into `components/_meta/_01-foot.twig` which feels silly. Seems like it would be great if we could read the same YAML file as Drupal instead of guessing at the dist directory. I'm going to look into that route.

evanmwillhite commented:
I have rewritten this to parse the same *.libraries.yml file - can I get an eye on the code there?

On the dependencies side, I am wavering on whether we should support autoloading that in Emulsify. What does everyone think?

evanmwillhite commented:
Thanks for reviewing this!! By autoloading, I mean autoloading dependencies. Basically, the libraries file can contain something like this:

```
main-menu:
  js:
    dist/02-molecules/menus/main-menu/main-menu.js: {}
  dependencies:
    - core/drupal
```

The function in Emulsify will only load `dist/02-molecules/menus/main-menu/main-menu.js` and not concern itself with dependencies. The advantage of this is that we don't have to figure out where those dependencies are relative to the emulsify installation. The disadvantage of this is that you have to manually add any symlinked dependencies in the [foot.twig](https://github.com/fourkitchens/emulsify/blob/develop/components/_meta/_01-foot.twig) file. If there is a clean way to figure out where dependencies are located, I'm OK with adding that to the script. If we can't do it cleanly or in a way that covers all situations, I'm also OK with leaving this manual process of adding symlinked deps in the foot.twig file. What do you think about that part?

evanmwillhite commented:
Sounds good to me. You want to green light this @ccjjmartin?

evanmwillhite commented:
Let's wait then - we should definitely functionally test this in PL and in a Drupal environment just to be sure. Thank you!

evanmwillhite commented:
I just thought about this over the weekend, and I think we should offer this separately over at https://github.com/drupal-pattern-lab, so I'm going to remove here and add over there and set emulsify up to load it automatically like the bem and add_attributes twig extensions. Moving back to in progress.

evanmwillhite commented:
OK, I've updated it and added the Twig extension here on [github](https://github.com/drupal-pattern-lab/attach-library-twig-extension) and [here](https://packagist.org/packages/drupal-pattern-lab/attach-library-twig-extension) on packagist. I've also updated the testing instructions above.
